---
title: "Dataset 2 RNA-Seq of pathogen inoculated arabidopsis with batch effects"
output: html_notebook
---

## This example is a remake of the example (4.2) presented in the edgeR documentation

```{r}
# Clear variables
rm(list=ls())
#Get path
path<-getwd(); path
# Set path
setwd(path)
```

Samples were collected in three batches, and adjustment for batch effects proves to be important. The aim of the analysis therefore is to detect genes differentially expressed (DE) in response to ΔhrcC challenge, while correcting for any differences between the batches.

Three biological replicates of the experiment were conducted at separate times and using independently grown plants and bacteria.

## Important libraries
```{r}
library(BiocGenerics)
library(BiocManager)
library(edgeR)
library(DESeq2)
library(ggplot2)
library(readr)
library(dplyr)
library(readxl)
library(foreign)
library(GO.db)
```



```{r}
arab <- read_rds("arab.rds")
head(arab)
```
There are two experimental factors, treatment (hrcc vs mock) and the time that each replicate was conducted ("1", "2" and "3"):

Treatment: 2 levels
Time: 3 levels

## Substring data

Substring: We can perform multiple things like extracting of values, replacement of values and more. For this we use functions like substr() and substring().

x = the input data / file
Start / First= starting index of the substring
Stop / Last= Ending index of the substring

substring(colnames(arab),1,4) this will get the first 4 characters of the column names, that is, h r c c and m o c k

while substring(colnames(arab),5,5) will get the fifth letter/character of the column names, that is, 1 2 3 1 2 3
```{r}
Treat <- factor(substring(colnames(arab),1,4))
Treat <- relevel(Treat, ref="mock")
# Define "mock" as the reference level

Time <- factor(substring(colnames(arab),5,5))
levels(Time) #Find the Time's levels order
```

Given that we alredy defined our variables as factors, we can then create a DGEList object:

```{r}
y <- DGEList(counts=arab, group=Treat)
```


## Filtering and normalization

There is no purpose in analysing genes that are not expressed in either experimental condition, so genes are first filtered on expression levels.

filterbYeXPR(): This function implements the filtering strategy that was described informally by Chen et al (2016). Roughly speaking, the strategy keeps genes that have at least min.count reads in a worthwhile number samples.

the filtering keeps genes that have CPM >= CPM.cutoff in MinSampleSize samples, where CPM.cutoff = min.count/median(lib.size)*1e6 and MinSampleSize is the smallest group sample size or, more generally, the minimum inverse leverage computed from the design matrix.

```{r}
keep <- filterByExpr(y)
table(keep)

y <- y[keep, , keep.lib.sizes=FALSE]
```

The TMM normalization is applied to account for the compositional biases:

```{r}
y <- normLibSizes(y)
y$samples
```

## Data exploration

An MDS plot shows the relative similarities of the six samples:
```{r}
plotMDS(y, col=rep(1:2, each=3))
```

HOW to interpret these results:

Distances on an MDS plot of a DGEList object correspond to leading log-fold-change between each pair of samples. Leading log-fold-change is the root-mean-square average of the largest log2-fold-changes between each pair of samples. Each pair of samples extracted at each time tend to cluster together, suggesting a batch effect. The hrcc treated samples tend to be below the mock samples for each time, suggesting a treatment effect within each time. The two samples at time 1 are less consistent than at times 2 and 3.


## predictive log2-foldchanges (logFC)

To examine further consistency of the three replicates, we compute predictive log2-foldchanges (logFC) for the treatment separately for the three times.

```{r}
design <- model.matrix(~Time+Time:Treat)
logFC <- predFC(y,design,prior.count=1,dispersion=0.05)

```



The logFC at the three times are positively correlated with one another, as we would hope:
(The correlation is highest between times 2 and 3)
```{r}
cor(logFC[,4:6])

cor(logFC[,1:3])
```


## Design matrix

Before we fit GLMs, we need to define our design matrix based on the experimental design.
We want to test for differential expressions between ΔhrcC challenged and mock-inoculated samples within batches, i.e. adjusting for differences between batches. In statistical terms, this is an additive linear model. So the design matrix is created as:


```{r}
design <- model.matrix(~Time+Treat)
rownames(design) <- colnames(y)
design
```

## Dispersion estimation


Estimate the genewise dispersion estimates over all genes, allowing for a possible abundance trend. The estimation is also robustified against potential outlier genes.

```{r}
y <- estimateDisp(y, design, robust=TRUE)
y$common.dispersion
```

```{r}
plotBCV(y)
```

PLOT INTERPRETATION:

The square root of dispersion is the coefficient of biological variation (BCV). The common BCV is on the high side, considering that this is a designed experiment using genetically identical plants. The trended dispersion shows a decreasing trend with expression level. 
At low logCPM, the dispersions are very large indeed.

Note that only the trended dispersion is used under the quasi-likelihood (QL) pipeline. The tagwise and common estimates are shown here but will not be used further.

## QL dispersions

The QL dispersions can be estimated using the glmQLFit function, and then be visualized with the plotQLDisp function.


```{r}
fit <- glmQLFit(y, design, robust=TRUE)
plotQLDisp(fit)
```

## Differential expression

Now we test for significant differential expression in each gene using the QL F-test.

(1) First we check whether there was a genuine need to adjust for the experimental times. We do this by testing for differential expression between the three times. There is considerable differential expression, justifying our decision to adjust for the batch effect:


```{r}
qlf <- glmQLFTest(fit, coef=2:3)
topTags(qlf)
```

```{r}
FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(FDR < 0.05)
```

(2) Now conduct QL F-tests for the pathogen effect and show the top genes. By default, the test is for the last coefficient in the design matrix, which in this case is the treatment effect:

```{r}
qlf <- glmQLFTest(fit)
topTags(qlf)
```


Here’s a closer look at the individual counts-per-million for the top genes. The top genes are very consistent across the three replicates:


```{r}
top <- rownames(topTags(qlf))
cpm(y)[top,]
```

The total number of genes significantly up-regulated or down-regulated at 5% FDR is summarized as follows:

```{r}
summary(decideTests(qlf))
```

We can plot all the logFCs against average count size, highlighting the DE (differentially expressed) genes:

(The blue lines indicate 2-fold up or down)
```{r}
plotMD(qlf)
abline(h=c(-1,1), col="blue")
```























